lib/CodeGen/MemoryAllocator.cpp

```diff
@@ -138,49 +154,573 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+       ++allocIt) {
+    if (!allocIt->alloc_) {
+      continue;
+    }
+    // Find a FREE instruction associated to this ALLOC.
+    auto allocHandle = allocIt->handle_;
+    bool freeFound = false;
+    for (auto it = allocIt; it != allocList.end(); ++it) {
+      if ((!it->alloc_) && (it->handle_ == allocHandle)) {
+        freeFound = true;
+        break;
+      }
+    }
+    if (!freeFound) {
+      return false;
+    }
+    // Verify ALLOC handle is unique.
+    auto handleIt =
+        std::find(allocHandleList.begin(), allocHandleList.end(), allocHandle);
+    if (handleIt != allocHandleList.end()) {
+      return false;
+    }
+    allocHandleList.push_back(allocHandle);
+  }
+
+  return true;
+}
+
+bool MemoryAllocator::verifySegments(
+    const std::list<Allocation> &allocList) const {
+
+  // Verify number of segments.
+  if (handleToSegmentMap_.size() != (allocList.size() / 2)) {
+    return false;
+  }
+
+  // Segments handles should match allocation handles.
+  // Segments sizes should match allocation sizes (with alignment).
+  for (const auto &alloc : allocList) {
+    if (!alloc.alloc_) {
+      continue;
+    }
+    auto it = handleToSegmentMap_.find(alloc.handle_);
+    if (it == handleToSegmentMap_.end()) {
+      return false;
+    }
+    Segment seg = it->second;
+    if (seg.size() != getEffectiveSize(alloc.size_)) {
+      return false;
+    }
+  }
+
+  // Allocations which are simultaneously alive must be assigned non-overlapping
+  // segments.
+  std::list<Handle> liveHandleList;
+  for (const auto &alloc : allocList) {
+    auto allocHandle = alloc.handle_;
+    if (alloc.alloc_) {
+      // Verify current segment is not overlapping with the other live segments.
+      auto it = handleToSegmentMap_.find(alloc.handle_);
+      Segment seg = it->second;
+      for (const auto &liveHandle : liveHandleList) {
+        auto liveIt = handleToSegmentMap_.find(liveHandle);
+        Segment liveSeg = liveIt->second;
+        bool segOverlap = intervalsOverlap(seg.begin_, seg.end_, liveSeg.begin_,
+                                           liveSeg.end_);
+        if (segOverlap) {
+          return false;
+        }
+      }
+      // Add handle to live handles.
+      liveHandleList.push_back(allocHandle);
+    } else {
+      // Remove handle from live handles.
+      auto it =
+          std::find(liveHandleList.begin(), liveHandleList.end(), allocHandle);
+      assert(it != liveHandleList.end() && "Handle not found for removal!");
+      liveHandleList.erase(it);
+    }
+  }
+  assert(liveHandleList.empty() && "Inconsistent allocations!");
+
+  return true;
+}
+
+/// Memory allocation strategy which defines the order used to allocate buffers.
+/// The strategies are listed in the decreasing order of the likelihood of
+/// getting the best allocation efficiency. These strategies are only used
+/// within the allocateAll function used to allocate all segments at once.
+enum class MemAllocStrategy : uint8_t {
```

Comment:
Have you considered allowing for custom memory allocation strategies? I.e. there are some supported by default, but users could register their own with a given instance of an allocator. Something like:

```c++
MemoryAllocator::MemoryAllocator(...) {
   /// Register standard strategies.
   registerMemAllocStrategy(MaxLiveSizeMaxBuffSizeImplClosure);
   ...
}
/// And then optionally later in user-code.
MemoryAllocator myAlloc(...);
/// Register my own mem alloc strategy implementation.
myAlloc.registerMemAllocStrategy(customMemAllocStrategyImplClosure);
```

Reply:
It is an interesting idea but I don't think this is necessary. The goal of the memory allocator is provide allocation with best efficiency (lowest memory usage). I don't think users would be interested in the details and in registering their own strategies.

Reply:
The users of allocators are e.g. Glow backends. Some backends may have very target specific allocation strategies, which should not be part of the general MemoryAllocator.

But I agree that it is a pretty advanced scenario and is probably not very important.

Reply:
For someone to register its own strategy would require some additional context (e.g. the mapping between segment IDs and handles). Since I don't have yet an example of how a user would want to specialize the allocation, I think it is best to leave this feature for later.